File: client/src/main/java/dev/quasemago/aura/client/infra/searches/dto/AnimeSearchResponseDTO.java
package dev.quasemago.aura.client.infra.searches.dto;

import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
public class AnimeSearchResponseDTO {
    private Pagination pagination;
    private List<Anime> data;

    @Getter
    @Setter
    public static class Pagination {
        private Integer last_visible_page;
        private Boolean has_next_page;
        private Integer current_page;
        private Items items;
    }

    @Getter
    @Setter
    public static class Items {
        private Integer count;
        private Integer total;
        private Integer per_page;
    }

    @Getter
    @Setter
    public static class Anime {
        private Integer mal_id;
        @JsonSetter(nulls = Nulls.SKIP)
        private String url = "N/A";
        private Images images;
        private Trailer trailer;
        private Boolean approved;
        private List<Title> titles;
        @JsonSetter(nulls = Nulls.SKIP)
        private String title = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String title_english = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String title_japanese = "N/A";
        private List<String> title_synonyms;
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String source = "N/A";
        private Integer episodes;
        @JsonSetter(nulls = Nulls.SKIP)
        private String status = "N/A";
        private Boolean airing;
        private Aired aired;
        @JsonSetter(nulls = Nulls.SKIP)
        private String duration = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String rating = "N/A";
        private Double score;
        private Integer scored_by;
        private Integer rank;
        private Integer popularity;
        private Integer members;
        private Integer favorites;
        @JsonSetter(nulls = Nulls.SKIP)
        private String synopsis = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String background = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String season = "N/A";
        private Integer year;
        private Broadcast broadcast;
        private List<Producer> producers;
        private List<Producer> licensors;
        private List<Producer> studios;
        private List<Genre> genres;
        private List<Theme> themes;
        private List<Demographic> demographics;
    }

    @Getter
    @Setter
    public static class Images {
        private ImageFormat jpg;
        private ImageFormat webp;
    }

    @Getter
    @Setter
    public static class ImageFormat {
        @JsonSetter(nulls = Nulls.SKIP)
        private String image_url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String small_image_url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String large_image_url = "N/A";
    }

    @Getter
    @Setter
    public static class Trailer {
        @JsonSetter(nulls = Nulls.SKIP)
        private String youtube_id = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String embed_url = "N/A";
        private TrailerImages images;
    }

    @Getter
    @Setter
    public static class TrailerImages {
        @JsonSetter(nulls = Nulls.SKIP)
        private String image_url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String small_image_url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String medium_image_url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String large_image_url = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String maximum_image_url = "N/A";
    }

    @Getter
    @Setter
    public static class Title {
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String title = "N/A";
    }

    @Getter
    @Setter
    public static class Aired {
        @JsonSetter(nulls = Nulls.SKIP)
        private String from = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String to = "N/A";
        private Prop prop;
        @JsonSetter(nulls = Nulls.SKIP)
        private String string = "N/A";
    }

    @Getter
    @Setter
    public static class Prop {
        private FromTo from;
        private FromTo to;
    }

    @Getter
    @Setter
    public static class FromTo {
        private Integer day;
        private Integer month;
        private Integer year;
    }

    @Getter
    @Setter
    public static class Broadcast {
        @JsonSetter(nulls = Nulls.SKIP)
        private String day = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String time = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String timezone = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String string = "N/A";
    }

    @Getter
    @Setter
    public static class Producer {
        @JsonSetter(nulls = Nulls.SKIP)
        private Integer mal_id = 0;
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String url = "N/A";
    }

    @Getter
    @Setter
    public static class Genre {
        private Integer mal_id;
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String url = "N/A";
    }

    @Getter
    @Setter
    public static class Theme {
        private Integer mal_id;
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String url = "N/A";
    }

    @Getter
    @Setter
    public static class Demographic {
        private Integer mal_id;
        @JsonSetter(nulls = Nulls.SKIP)
        private String type = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String name = "N/A";
        @JsonSetter(nulls = Nulls.SKIP)
        private String url = "N/A";
    }
}

File: src/infrastructure/jikan/dtos/anime-search-response-dto.ts
import { Type } from "class-transformer";
import {
  IsArray,
  IsBoolean,
  IsInt,
  IsNumber,
  IsOptional,
  IsString,
  ValidateNested
} from "class-validator";

export class AnimeSearchResponseDTO {
  @ValidateNested()
  @Type(() => Pagination)
  pagination?: string;

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Anime)
  data?: Anime[];

  @IsString()
  createdBy?: string;
}

export class Pagination {
  @IsInt()
  last_visible_page!: number;

  @IsBoolean()
  has_next_page!: boolean;

  @IsInt()
  current_page!: number;

  @ValidateNested()
  @Type(() => Items)
  items!: Items;
}

export class Items {
  @IsInt()
  count!: number;

  @IsInt()
  total!: number;

  @IsInt()
  per_page!: number;
}

export class Anime {
  @IsInt()
  mal_id!: number;

  @IsString()
  url?: string;

  @ValidateNested()
  @Type(() => Images)
  images?: Images;

  @ValidateNested()
  @Type(() => Trailer)
  trailer?: Trailer;

  @IsBoolean()
  approved?: boolean;

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Title)
  titles?: Title[];

  @IsString()
  title?: string;

  @IsString()
  title_english?: string;

  @IsString()
  title_japanese?: string;

  @IsArray()
  title_synonyms?: string[];

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsString()
  @IsOptional()
  source: string = "N/A";

  @IsInt()
  episodes?: number;

  @IsString()
  @IsOptional()
  status: string = "N/A";

  @IsBoolean()
  airing?: boolean;

  @ValidateNested()
  @Type(() => Aired)
  aired?: Aired;

  @IsString()
  @IsOptional()
  duration: string = "N/A";

  @IsString()
  @IsOptional()
  rating: string = "N/A";

  @IsNumber()
  score?: number;

  @IsInt()
  scored_by?: number;

  @IsInt()
  rank?: number;

  @IsInt()
  popularity?: number;

  @IsInt()
  members?: number;

  @IsInt()
  favorites?: number;

  @IsString()
  @IsOptional()
  synopsis: string = "N/A";

  @IsString()
  @IsOptional()
  background: string = "N/A";

  @IsString()
  @IsOptional()
  season: string = "N/A";

  @IsInt()
  year?: number;

  @ValidateNested()
  @Type(() => Broadcast)
  broadcast?: Broadcast;

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Producer)
  producers?: Producer[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Producer)
  licensors?: Producer[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Producer)
  studios?: Producer[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Genre)
  genres?: Genre[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Theme)
  themes?: Theme[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Demographic)
  demographics?: Demographic[];
}

export class Images {
  @ValidateNested()
  @Type(() => ImageFormat)
  jpg?: ImageFormat;

  @ValidateNested()
  @Type(() => ImageFormat)
  webp?: ImageFormat;
}

export class ImageFormat {
  @IsString()
  image_url?: string;

  @IsString()
  small_image_url?: string;

  @IsString()
  large_image_url?: string;
}

export class Trailer {
  @IsString()
  @IsOptional()
  youtube_id: string = "N/A";

  @IsString()
  url?: string;

  @IsString()
  embed_url?: string;

  @ValidateNested()
  @Type(() => TrailerImages)
  images?: TrailerImages;
}

export class TrailerImages {
  @IsString() image_url?: string;
  @IsString() small_image_url?: string;
  @IsString() medium_image_url?: string;
  @IsString() large_image_url?: string;
  @IsString() maximum_image_url?: string;
}

export class Title {
  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsString()
  @IsOptional()
  title: string = "N/A";
}

export class Aired {
  @IsString()
  @IsOptional()
  from: string = "N/A";

  @IsString()
  @IsOptional()
  to: string = "N/A";

  @ValidateNested()
  @Type(() => Prop)
  prop?: Prop;

  @IsString()
  @IsOptional()
  string: string = "N/A";
}

export class Prop {
  @ValidateNested()
  @Type(() => FromTo)
  from?: FromTo;

  @ValidateNested()
  @Type(() => FromTo)
  to?: FromTo;
}

export class FromTo {
  @IsInt() day?: number;
  @IsInt() month?: number;
  @IsInt() year?: number;
}

export class Broadcast {
  @IsString()
  @IsOptional()
  day: string = "N/A";

  @IsString()
  @IsOptional()
  time: string = "N/A";

  @IsString()
  @IsOptional()
  timezone: string = "N/A";

  @IsString()
  @IsOptional()
  string: string = "N/A";
}

export class Producer {
  @IsInt()
  @IsOptional()
  mal_id: number = 0;

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsString()
  @IsOptional()
  name: string = "N/A";

  @IsString()
  url?: string;
}

export class Genre {
  @IsInt() mal_id?: number;

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsString()
  @IsOptional()
  name: string = "N/A";

  @IsString()
  url?: string;
}

export class Theme {
  @IsInt() mal_id?: number;

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsString()
  @IsOptional()
  name: string = "N/A";

  @IsString()
  url?: string;
}

export class Demographic {
  @IsInt() mal_id?: number;

  @IsString()
  @IsOptional()
  type: string = "N/A";

  @IsString()
  @IsOptional()
  name: string = "N/A";

  @IsString()
  url?: string;
}
